implemented Detailedlist for Gtk

elelay opened the pull request:
This is a first implementation of the DetailedList for gtk.

Tests will be added, but I'd like a first feedback.

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

elelay commented:
Thanks for the feedback.

Regarding abstracting row rendering, I've submitted a new IconTextRenderer doing a fair job of rendering combinations of *title*, *title with subtitle*, *icon*, with named styles like *title-icon* for icon on the right, but please share what you have in mind for customization, because to my mind it should be implemented in the core widget...

Regarding the model, I'll look into this now.

Regarding testing, I've tried moving 'widgets/detailedlist.py' from mobile only to desktop and all tests pass. But it shouldn't be committed until all desktop implementations support it, right?

elelay commented:
@freakboy3742 here is the SourcetreeModel. It needed a few additions to source notifications:
 - add index (and parent for TreeSource) of the item to the **remove** notification
 - add old_data to the **clear** notification

Do you agree?

I didn't encounter any exception when selecting rows.

I understand that I should copy and adapt existing tests of tree and table gtk implementations for the detailedlist. it'll come next week.

freakboy3742 commented:
Adding the index to the remove notification makes sense to me - I can see how it could be needed to remove based on list index. However, I'm not sure why the `old_data` is needed on `clear`... don't we just need a "burn everything" option in that case?

elelay commented:
>  However, I'm not sure why the old_data is needed on clear... don't we just need a "burn everything" option in that case?
There is no such api on the TreeModel, so it must [notify the tree of every removal](https://gitlab.gnome.org/GNOME/gtk/blob/3.24.18/gtk/gtktreestore.c#L1883).

freakboy3742 commented:
Sure - but surely you can get access to all the *current* nodes on the tree, and delete them. It makes sense that there might not be a "clear all" API, but the model itself *should* have a "get me everything I contain" API, which is the deletion list you need. It shouldn't need to be passed in as an argument from a separate data store.

elelay commented:
What I meant by *there is no such api on the TreeModel* is that it doesn't have a signal to tell the TreeView that it's empty or changed dramatically (that would be convenient), so the TreeModel must notify the TreeView that each node is removed (leafs first).

> Sure - but surely you can get access to all the _current_ nodes on the tree, and delete them.

Indeed, in gtk, the TreeModel is responsible for storing all the _current_ nodes. What I perhaps didn't make clear is that SourceTreeModel doesn't store any structure, only references to all the nodes in a dict, with their id as the key, this only to store a reference to the node in the Gtk.TreeIter native object.

> It shouldn't need to be passed in as an argument from a separate data store

My design goal was to store as little as possible, going back to the TreeSource or ListSource to do all the traversal, get the index of the node, etc..

I really need some guidance in what you want for the model implementation.

An example trade-off I made: navigation (going to the next node) is a bit slow, because the node's index is not stored in the Gtk.TreeIter, needing me to search it and it's parents in arrays in the TreeSource. That's not great, but what's great is that Gtk.TreeIter don't become invalid until the node they point to is deleted (not invalidated by the insertion/deletion of a node before them. So I can advertise that stability to a sorted model or the treeview, which have optimizations for that.

What I could have done would be to store the index of each node in its parent. This way, I can quickly go to the next or previous node in the parent and quickly construct the indices to form the TreePath. But each time I insert/delete a node I need to iterate through its following siblings (in the Source) and increase/decrease their index. Now it seems a better idea overall.

Hope this makes sense...

freakboy3742 commented:
> Hope this makes sense...

Not entirely. I understand that the intent is not to store structure on the TreeModel (and that's what I would have expected); however, what I don't understand is why "stop showing *everything*" requires a "give me a list of everything you want to not show" argument. 

The TreeModel is tracking a list of GTK-level nodes that it knows about - I don't see why *that* isn't already a list of everything that needs to be removed. Sure - it's a dictionary rather than a list, but surely it's still an iterable of all visible content?

Another approach would be to *completely* replace the tree model. When the TreeView's source is changed, you're already setting the model to None and then back again (presumably to avoid bulk creating data modification signals). Is there any reason that you can't create *an entirely new model*, instead of clearing and re-using the existing model - and do the same for clear? Why go to great lengths to reuse a single Model instance when you can create an entirely new, clean model?

As for the implementation being slow because of the need to do linear lookup - I'm not at all clear why that's the case. I'm not intimately familar with GTK TreeModel, but it looks like it's iteration based, rather than index based. ListSource and TreeSource support both index lookup and `__iter__`... so what's the complication? 

To be clear - I'm entirely open to the idea of modifying the API of Source to meet the requirements of GTK - after all, it's one of the platforms we support. It's just not clear to me at present what the GTK API requirement *is* that the Source API doesn't currently satisfy.

elelay commented:
> Another approach would be to completely replace the tree model. When the TreeView's source is changed, you're already setting the model to None and then back again (presumably to avoid bulk creating data modification signals). Is there any reason that you can't create an entirely new model, instead of clearing and re-using the existing model - and do the same for clear? Why go to great lengths to reuse a single Model instance when you can create an entirely new, clean model?

Yes, that's a really good suggestion: I was so focused on mapping Source changes to TreeModel events that I didn't consider implementing the clear() operation within the detailedlist, tree and table.
I'll push something to do that soon.

elelay commented:
Hi @freakboy3742,
I've now modified the sourcetreemodel to store item's index in parent in the model to improve performance.
Also removed the `old_data` parameter on the `clear()` callback.

Can you review again the PR?
Thanks,

elelay commented on examples/detailedlist/detailedlist/app.py:
```diff
@@ -10,7 +10,10 @@
+        self.label.text = 'Bee is {} in {}'.format(row.title, row.subtitle) \
+            if row is not None else 'No row selected'
+        # work around no selection property in DetailedList as in Tree by keeping the last selected row
+        self.selected_row = row
```
Agreed, it was a not so subtle reminder to report #962...

elelay replied:
I wanted to test deletion so needed the selection. I'll add a better comment with reference to the issue and TODO remove

elelay commented on src/cocoa/toga_cocoa/widgets/table.py:
```diff
@@ -188,7 +188,7 @@ def change(self, item):
+    def remove(self, item, index):
         try:
             # We can't get the index from self.interface.data because the
             # row has already been removed. Instead we look up the index
```
Is it enough to add a todo here? I'm not familiar with the cocoa implementation...

elelay commented on src/core/toga/sources/tree_source.py:
```diff
@@ -134,12 +138,15 @@ def append(self, parent, *values, **named):
         else:
-            node._parent._children.remove(node)
+            del node._parent._children[i]
+            del node._parent
```
To help gc, but the cycle is already broken by deleting node from the parent.

elelay replied:
Well, it's more an invariant thing: node is not in parent's children so I prefer that its _parent attribute is also cleared.

elelay commented:
Thanks for the review!
I think I've addressed your remarks, except for the rendering issue, which I can't reproduce in my environment (ArchLinux, gtk 3.24).
Here **Chinese (Traditional)** or **Chinese (Simplified** are not cropped.

freakboy3742 reviewed (approved):
Apologies for the delays on this; I took a look at this a couple of weeks ago, and then life got complicated.

From the perspective of your code, I think this is good to go. The inconsistent font-size thing with Chinese would be nice to fix, but what you've contributed here is a significant improvement over the status quo, so I don't think that fix needs to stand in the way of this being merged. 

However, there's some macOS-specific stuff that needs to be looked at - in particular, the arguments to the `selected` event handler have been modified fairly significantly altered. I agreed to take a look at the macOS parts a while back; I just need to find some time...

freakboy3742 reviewed (changes requested):
Ok - I've made the updates for Cocoa (and also iOS and Android); however, in doing some final testing, I need to roll back my previous approval.

On GTK, the detailed list, table, table_source and tree demos all work fine - but tree_source fails. It looks like there's something odd going on with the way the new GTK data source is handling the root node. Are you able to dig into what is going on there?

elelay commented:
Great that you could pick it up! Honestly I was afraid it would be left to bitrot.

Indeed, sourcetreemodel expects `row._parent` to be `None` for roots. This is not the case in the `FileSystemSource` in `tree_source` example. I'm fixing this now.

Also an issue with icon + text as the *Name* column. I'll look into it.

elelay commented:
I can't find a way to implement icon+text in a tree view when attr returns a tuple. I can strip the first value of any tuple in the sourcetreemodel's do_get_value.

freakboy3742 commented:
The absence of icons isn't a huge problem. The old code didn't support icons, and while it would be nice to have icons, I don't consider them a blocker. However, the current code is currently eating the *text* from the label when an icon has been provided. 

I've also noticed some problems around selection handling. If you get a chance to look at this, that would be really helpful; if you don't, I'll try and take a look later this week (or this weekend).

elelay commented:
Eaten text is now fixed (was caused by trying to affect a tuple to the native text property of the cell renderer) with a workaround in the sourcetreemodel.

Also fixed weird selection counts: it was caused by returning directly the Node, and using its len (ie how many files in the folder) as the selection len.

freakboy3742 reviewed (approved):
Ok - I've made a couple of small tweaks, but I think this is good to go. It's not perfect (icons, multi-select, and double click aren't working, plus the rendering problem that I reported with Chinese) - but they weren't working before either, and this is significantly better than what we had. I'd also like to address the way we're handling selection so that the widget is a true source of truth, rather than using a copy of the data that is maintained on Toga's end - but that's something we can tackle as an incremental improvement. 

On that basis, I'm happy to merge! Thanks for all your work on this, and apologies for the delay in the review process!

elelay commented:
Great! Thanks for the help and review: we made it better together :-)
